crates/types/src/string.rs

```diff
@@ -115,6 +129,95 @@ impl String {
     }
 }
 
+impl StringBuilder {
+    /// Create a new empty `StringBuilder`.
+    #[inline]
+    pub fn new() -> Self {
+        Self { data: ptr::null_mut(), len: 0, size: 0 }
+    }
+
+    /// Push new bytes to the builder.
+    #[inline]
+    pub fn push_bytes(&mut self, bytes: &[u8]) {
+        if self.data.is_null() {
+            let len = bytes.len();
+            let size = len + 1;
+
+            let data = unsafe {
+                let data = libc::malloc(size) as *mut ffi::c_char;
+
+                libc::memcpy(
+                    data as *mut _,
+                    bytes.as_ptr() as *const _,
+                    bytes.len(),
+                );
+
+                *data.add(len) = 0;
+
+                data
+            };
+
+            self.data = data;
+            self.len = len;
+            self.size = size;
+
+            return;
+        }
+
+        let slice_len = bytes.len();
+        let required_size = self.len + slice_len + 1;
+
+        // Reallocate if pushing the bytes overflows the allocated memory.
+        if self.size < required_size {
+            // Same as `let n = (required_size as f64 / self.size as f64).ceil()` but using integer
+            // arithmetic.
+            //
+            // `n` is the smallest number that, when multiplied by `self.size`, gives a
+            // number that is at least equal to `required_size`.
+            let n = (required_size + self.size - 1) / self.size;
+            let new_size = self.size * n;
+
+            self.data = unsafe {
+                libc::realloc(self.data as *mut _, new_size)
+                    as *mut ffi::c_char
+            };
+
+            self.size = new_size;
+            debug_assert!(self.len < self.size)
```

Comment:
I think this strategy is a bit too conservative, and might lead to unnecessary re-allocations. `Vec`'s strategy is to just double the capacity until it's greater than the new length.

Reply:
What I meant is that we always keep the capacity a multiple of 2. It starts not as `len + 1`, but as the smallest power of 2 that's greater than the slice's length (with a minumum of 4). To get the new capacity, we just:

```rust
let n = (required_cap - 1).ilog2() + 1;
let new_cap = 2_usize.pow(n);
```

or even:

```rust
loop {
    self.cap *= 2;
    if required_cap <= self.cap {
        break;
    }
}
```